common/certificate/src/device_id.rs

```diff
@@ -0,0 +1,126 @@
+fn is_valid_device_id_char(ch: char) -> bool {
+    matches!(ch, 'A'..='Z'
+    | 'a'..='z'
+    | '0'..='9'
+    | '\''
+    | '='
+    | '('
+    | ')'
+    | '+'
+    | ','
+    | '-'
+    | '.'
+    | ':'
+    | '?'
+    | '_'
+    | '#'
+    | '*'
+    | '@'
+    | '!'
+    | '%')
+}
+
+pub fn is_valid_device_id(id: &str, max_cn_size: usize) -> Result<(), DeviceIdError> {
+    if id.is_empty() {
+        return Err(DeviceIdError::EmptyName);
+    } else if id.len() > max_cn_size {
+        return Err(DeviceIdError::TooLongName {
+            name: id.into(),
+            max_cn_size,
+        });
+    } else if !id.chars().all(is_valid_device_id_char) {
+        return Err(DeviceIdError::InvalidCharacter { name: id.into() });
+    }
+
```

Comment:
@itsyitsy @didier-wenzek What about single character names? I feel those shouldn't be allowed either especially since we allow special characters like `+` and `#` which are "really special" in the MQTT world. If someone names their device `+` or `#` and if that appears in the topic names, the mosquitto broker or even the cloud broker may not take it that well.

Reply:
I would not bother too much with a minimum id size, the main point being that the device ids are unique at the tenant level. However, it might be better to reject `+` and `#` that are MQTT reserved words. This will let us use child device ids as topic names.